source/common/stats/thread_local_store.h

```diff
@@ -171,19 +171,33 @@ class ThreadLocalStoreImpl : Logger::Loggable<Logger::Id::stats>, public StoreRo
   const Stats::StatsOptions& statsOptions() const override { return stats_options_; }
 
 private:
-  template <class Stat> using StatMap = CharStarHashMap<Stat>;
+  friend class ThreadLocalStoreTestScope;
+
+  template <class Stat> using StatMap = ConstCharStarHashMap<Stat>;
+  using SharedString = std::shared_ptr<std::string>;
+  using SharedStringSet =
+      absl::flat_hash_map<const char*, SharedString, ConstCharStarHash, ConstCharStarEqual>;
```

Comment:
Do we actually need a map here? Can this just be a set of SharedString with the right hash/equality function?

Reply:
Last night I had typed in a comment but hadn't committed it yet:
```
+  // We use an absl::flat_hash_map to implement what's conceptually a set of
+  // shared strings.  This structure enables us to do a find() without
+  // constructing a shared_string, which would entail making a full copy of the
+  // stat name.
```

Reply:
Cool thanks for the explanation. According to the docs flat_hash_map supports heterogeneous lookup. Could we use that to avoid the copy?

Reply:
Yes! It took me a little while to figure out how to do that, but done.